// simple-react-form/lib/Contexts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParentFieldNameContext = exports.OnChangeContext = exports.ErrorMessagesContext = exports.ValueContext = exports.useFormOnChangeContext = exports.useErrorMessagesContext = exports.useFormParentFieldNameContext = exports.useFormValueContext = void 0;
const tslib_1 = require("tslib");
const react_1 = (0, tslib_1.__importStar)(require("react"));
const ValueContext = react_1.default.createContext({});
exports.ValueContext = ValueContext;
const ParentFieldNameContext = react_1.default.createContext(null);
exports.ParentFieldNameContext = ParentFieldNameContext;
const ErrorMessagesContext = react_1.default.createContext({});
exports.ErrorMessagesContext = ErrorMessagesContext;
const OnChangeContext = react_1.default.createContext((fieldName, fieldValue) => { });
exports.OnChangeContext = OnChangeContext;
function useFormValueContext() {
    return (0, react_1.useContext)(ValueContext);
}
exports.useFormValueContext = useFormValueContext;
function useFormParentFieldNameContext() {
    return (0, react_1.useContext)(ParentFieldNameContext);
}
exports.useFormParentFieldNameContext = useFormParentFieldNameContext;
function useErrorMessagesContext() {
    return (0, react_1.useContext)(ErrorMessagesContext);
}
exports.useErrorMessagesContext = useErrorMessagesContext;
function useFormOnChangeContext() {
    return (0, react_1.useContext)(OnChangeContext);
}
exports.useFormOnChangeContext = useFormOnChangeContext;
